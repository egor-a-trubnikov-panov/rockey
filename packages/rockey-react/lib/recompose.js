import React from 'react';
import isString from 'lodash/isString';
import isFunction from 'lodash/isFunction';
import compose from 'recompose/compose';

import filterProps from './utils/filterProps';
import htmlTags from './htmlTags';
import rockey from './';

const ucfirst = s => s.charAt(0).toUpperCase() + s.slice(1);

let anonCounter = 0;

const RecomposeRockeyHoc = (...args) => {
  let displayName = null;
  let BaseComponent = null;

  // TODO: check types
  if (args.length === 1) {
    displayName = `RockeyRecompose${++anonCounter}`;
    BaseComponent = args[0];
  } else {
    displayName = args[0];
    BaseComponent = args[1];
  }

  return (...enhancers) => {
    const RecomposeHoc = compose(...enhancers)(BaseComponent);

    return rockey(ucfirst(displayName), RecomposeHoc);
  };
};

let counter = {};

for (const tag of htmlTags) {
  // ---- tag hoc lazy creation
  Object.defineProperty(RecomposeRockeyHoc, tag, {
    get: () => (displayName, ...args) => {
      let name = null;
      let enhancers = null;

      if (isString(displayName)) {
        name = ucfirst(displayName);
        enhancers = args;
      } else if (isFunction(displayName)) {
        if (!counter[tag]) {
          counter[tag] = 0;
        }

        name = `RecomposeShortcut${ucfirst(tag)}${++counter[tag]}`;
        enhancers = [displayName, ...args];
      } else {
        throw new Error(`wrong usage`);
      }

      const TagComponent = props => {
        return React.createElement(tag, filterProps(props));
      };

      return RecomposeRockeyHoc(name, TagComponent)(...enhancers);
    },
    set: () => {
      throw new Error('Override default RockeyHoc shorcuts is disabled');
    },
  });
}

export default RecomposeRockeyHoc;
