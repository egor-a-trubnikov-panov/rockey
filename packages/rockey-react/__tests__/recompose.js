import React from 'react';
import 'react-dom';
import renderer from 'react-test-renderer';
import { withProps, mapProps } from 'recompose';

import rockempose from '../lib/recompose';

jest.mock('rockey/utils/hash', () => {
  return () => '-hash-';
});

test('Render recompose', () => {
  let i = 0;

  const Button = rockempose.button(
    withProps(props => ({
      bg: 'yellow',
    })),
    mapProps(props => ({
      ...props,
      c: props.bg,
    }))
  )`
    color: ${props => props.c}
    bg: ${props => props.bg}
  `;

  const ButtonTree = renderer.create(<Button>Button</Button>).toJSON();
  expect(ButtonTree).toMatchSnapshot();
});

test('Render recompose with name', () => {
  let i = 0;

  const Button = rockempose.button(
    'MyButtonYo',
    withProps(props => ({
      bg: 'yellow',
    })),
    mapProps(props => ({
      ...props,
      c: props.bg,
    }))
  )`
    color: ${props => props.c}
    bg: ${props => props.bg}
  `;

  const ButtonTree = renderer.create(<Button>Button</Button>).toJSON();
  expect(ButtonTree).toMatchSnapshot();
});

test('Render recompose custom component', () => {
  let i = 0;

  const Button = rockempose(({ children, ...props }) => (
    <button {...props}>{children}</button>
  ))(
    withProps(props => ({
      bg: 'yellow',
    })),
    mapProps(props => ({
      ...props,
      c: props.bg,
    }))
  )`
    color: ${props => props.c}
    bg: ${props => props.bg}
  `;

  const ButtonTree = renderer.create(<Button>Button</Button>).toJSON();
  expect(ButtonTree).toMatchSnapshot();
});
