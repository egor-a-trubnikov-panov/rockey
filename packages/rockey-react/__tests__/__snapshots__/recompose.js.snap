// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render recompose 1`] = `
<button
  className="RecomposeShortcutButton1--hash-"
>
  Button
</button>
`;

exports[`Render recompose custom component 1`] = `
<button
  bg="yellow"
  c="yellow"
  className="RockeyRecompose1--hash-"
>
  Button
</button>
`;

exports[`Render recompose with name 1`] = `
<button
  className="MyButtonYo--hash-"
>
  Button
</button>
`;
